src/query/parallel/px_hash_join.cpp

```diff
@@ -0,0 +1,985 @@
+    {
+      QFILE_LIST_ID *list_id;
+      QFILE_LIST_ID **part_list_id;
+      QFILE_LIST_ID **temp_part_list_id = nullptr;
+
+      PAGE_PTR page = nullptr;
+      QFILE_TUPLE_RECORD tuple_record = { nullptr, 0 };
+      int tuple_cnt, tuple_index, tuple_length, tuple_offset;
+
+      VPID overflow_vpid = VPID_INITIALIZER;
+      PAGE_PTR overflow_page = nullptr;
+      QFILE_TUPLE_RECORD overflow_record = { nullptr, 0 };
+      int copy_offset, copy_size;
+
+      HASH_SCAN_KEY *temp_key = nullptr;
+      unsigned int hash_key;
+      UINT32 part_cnt, part_index, part_id;
+
+      bool is_outer_join = false;
+      bool need_skip_next = false;
+
+      int error = NO_ERROR;
+      bool has_error = false;
+
+      /* Do not perform NULL checks;
+       * validation is expected to be handled by the constructor */
+      list_id = m_split_info->fetch_info->list_id;
+      part_list_id = m_split_info->part_list_id;
+      part_cnt = m_manager->context_cnt;
+
+      is_outer_join = IS_OUTER_JOIN_TYPE (m_manager->join_type);
+
+      temp_part_list_id = (QFILE_LIST_ID **) db_private_alloc (&thread_ref, part_cnt * sizeof (QFILE_LIST_ID *));
+      if (temp_part_list_id == nullptr)
+	{
+	  assert_release (er_errid () != NO_ERROR);
+	  m_task_manager.handle_error (thread_ref);
+	  return;
+	}
+      memset (temp_part_list_id, 0, part_cnt * sizeof (QFILE_LIST_ID *));
+
+      temp_key = qdata_alloc_hscan_key (&thread_ref, m_manager->key_cnt, true);
+      if (temp_key == nullptr)
+	{
+	  assert_release (er_errid () != NO_ERROR);
+	  m_task_manager.handle_error (thread_ref);
+
+	  /* cleanup */
+	  db_private_free_and_init (&thread_ref, temp_part_list_id);
+
+	  return;
+	}
+
+      /* next page */
+      do
+	{
+	  if (m_task_manager.has_error () || m_task_manager.check_interrupt (thread_ref))
+	    {
+	      has_error = true;
+	      break;
+	    }
+
+	  page = get_next_page (thread_ref);
+	  if (page == nullptr)
+	    {
+	      if (er_errid () != NO_ERROR)
+		{
+		  m_task_manager.handle_error (thread_ref);
+		  has_error = true;
+		}
+
+	      /* end */
+	      break;
+	    }
+
+	  tuple_record.tpl = page;
+
+	  tuple_cnt = QFILE_GET_TUPLE_COUNT (page);
+	  if (tuple_cnt == 0)
+	    {
+	      /* empty page */
+	      continue;
+	    }
+	  tuple_index = -1;
+	  tuple_offset = 0;
+
+	  /* next tuple */
+	  do
+	    {
+	      if (tuple_index == -1)
+		{
+		  /* first tuple */
+		  tuple_offset = QFILE_PAGE_HEADER_SIZE;
+		  tuple_record.tpl = (char *) page + QFILE_PAGE_HEADER_SIZE;
+		}
+	      else if (tuple_index < tuple_cnt - 1)
+		{
+		  /* next tuple */
+		  tuple_length = QFILE_GET_TUPLE_LENGTH (tuple_record.tpl);
+		  tuple_offset += tuple_length;
+		  tuple_record.tpl += tuple_length;
+		}
+	      else
+		{
+		  /* next page */
+		  assert (tuple_index == tuple_cnt - 1);
+		  break;
+		}
+
+	      tuple_index++;
+
+	      /* overflow page */
+	      if (QFILE_GET_OVERFLOW_PAGE_ID (page) != NULL_PAGEID)
```

Comment:
```suggestion
	      if (__builtin_expect(QFILE_GET_OVERFLOW_PAGE_ID (page) != NULL_PAGEID, 0))
```
split 시 row-by-row로 overflow page가 있는지 확인한다면 성능 저하가 생길까 염려됩니다. 이 부분에서 성능 부하가 적나요? 또한, `get_next_page (thread_ref)` 이후에 overflow page가 존재하는 페이지라면 아예 다른 루틴을 통해 iteration 하는 것도 최적화할 수 있는 방법이 될 것 같습니다.

Reply:
따로 의견을 주신대로 오버플로우 페이지에 대한 분기를 do { ... } while 문 밖으로 분리하는 것으로 해결했습니다. (
d45156f)